crates/core/src/equivalence.rs

```diff
@@ -9,41 +10,38 @@ impl<'a> Binding<'a> {
     pub fn is_equivalent_to(&self, other: &'a Binding) -> bool {
         // covers Node, and List with one element
         if let (Some(s1), Some(s2)) = (self.singleton(), other.singleton()) {
-            return are_equivalent(s1.source, &s1.node, s2.source, &s2.node);
+            return are_equivalent(&s1, &s2);
         }
 
         match self {
             // should never occur covered by singleton
-            Self::Node(source1, node1) => match other {
-                Self::Node(source2, node2) => are_equivalent(source1, node1, source2, node2),
+            Self::Node(node1) => match other {
+                Self::Node(node2) => are_equivalent(node1, node2),
                 Self::String(str, range) => {
                     str[range.start_byte as usize..range.end_byte as usize] == self.text()
                 }
                 Self::FileName(_) | Self::List(..) | Self::Empty(..) | Self::ConstantRef(_) => {
                     false
                 }
             },
-            Self::List(source1, parent_node1, field1) => match other {
-                Self::List(source2, parent_node2, field2) => {
-                    let mut cursor1 = parent_node1.walk();
-                    let mut cursor2 = parent_node2.walk();
-                    children_by_field_id_count(parent_node1, *field1)
-                        == children_by_field_id_count(parent_node2, *field2)
-                        && parent_node1
-                            .children_by_field_id(*field1, &mut cursor1)
-                            .zip(parent_node2.children_by_field_id(*field2, &mut cursor2))
-                            .all(|(node1, node2)| are_equivalent(source1, &node1, source2, &node2))
-                }
+            Self::List(parent_node1, field1) => match other {
+                Self::List(parent_node2, field2) => parent_node1
+                    .children_by_field_id(*field1)
+                    .zip_longest(parent_node2.children_by_field_id(*field2))
+                    .all(|zipped| match zipped {
+                        EitherOrBoth::Both(node1, node2) => are_equivalent(&node1, &node2),
+                        EitherOrBoth::Left(_) | EitherOrBoth::Right(_) => false,
+                    }),
```

Comment:
Before, we compared `children_by_field_id_count()` of both nodes in a separate step, which meant we would often end up iterating the children twice: Once for counting, and once for comparing. Using `zip_longest()`, we can do both in one iteration.